core/chains/evm/logpoller/log_poller.go

```diff
@@ -398,14 +503,26 @@ func (lp *logPoller) pollAndSaveLogs(ctx context.Context, currentBlockNumber int
 			return
 		}
 		lp.lggr.Infow("Unfinalized log query", "logs", len(logs), "currentBlockNumber", currentBlockNumber, "blockHash", currentBlock.Hash())
+		logsByEventID := lp.groupByEventID(logs)
 		err = lp.orm.q.WithOpts(pg.WithParentCtx(ctx)).Transaction(func(tx pg.Queryer) error {
 			if err3 := lp.orm.InsertBlock(currentBlock.Hash(), currentBlockNumber, pg.WithQueryer(tx)); err3 != nil {
 				return err3
 			}
 			if len(logs) == 0 {
 				return nil
 			}
-			return lp.orm.InsertLogs(convertLogs(lp.ec.ChainID(), logs), pg.WithQueryer(tx))
+			if err3 := lp.orm.InsertLogs(convertLogs(lp.ec.ChainID(), logs), pg.WithQueryer(tx)); err3 != nil {
+				return err3
+			}
+			for eventID, logsForEventID := range logsByEventID {
+				if _, ok := lp.callbacks[eventID]; !ok {
+					continue
+				}
+				if err3 := lp.callbacks[eventID].Save(tx, logsForEventID); err3 != nil {
+					return err3
```

Comment:
Since we are _five_ indents and _`err3`_, maybe consider a helper func/method instead to return this anonymous `func(pg.Queryer) error`? :shrug: